tap_google_search_console/helpers.py

```diff
@@ -0,0 +1,164 @@
+
+
+def encode_and_format_url(url: str, string_format: str) -> str:
+    """
+    Encodes the given site_url
+    """
+    return string_format.format(quote(url, safe=""))
+
+
+def convert(name):
+    """
+    Converts a CamelCased word to snake case
+    """
+    regsub = re.sub('(.)([A-Z][a-z]+)', r'\1_\2', name)
+    return re.sub('([a-z0-9])([A-Z])', r'\1_\2', regsub).lower()
+
+
+# Convert keys in json array
+def convert_array(arr):
+    """
+    Converts all the CamelCased dict Keys in a list to snake case
+    Iterated through each object recursively and if the object is dict type then converts its key to snake case
+    """
+    new_arr = []
+    for element in arr:
+        if isinstance(element, list):
+            new_arr.append(convert_array(element))
+        elif isinstance(element, dict):
+            new_arr.append(convert_json(element))
+        else:
+            new_arr.append(element)
+    return new_arr
+
+
+# Convert keys in json
+def convert_json(this_json):
+    """
+    Converts all the CamelCased Keys in a nested dictionary object to snake case
+    """
+    out = {}
+    for key in this_json:
+        new_key = convert(key)
+        if isinstance(this_json[key], dict):
+            out[new_key] = convert_json(this_json[key])
+        elif isinstance(this_json[key], list):
+            out[new_key] = convert_array(this_json[key])
+        else:
+            out[new_key] = this_json[key]
+    return out
+
+
+
+def remove_keys_nodes(this_json, path):
+    """
+    Remove `keys` node, if exists
+    """
+    new_json = this_json
+    idx = 0
+    for record in list(new_json[path]):
+        if record.get("keys", None):
+            new_json[path][idx].pop("keys")
+        idx = idx + 1
+    return new_json
+
+
+def hash_data(data):
+    """
+    Create MD5 hash key for data element
+    Prepares the project id hash
+    """
+    hash_id = hashlib.md5()
+    hash_id.update(repr(data).encode("utf-8"))
+    return hash_id.hexdigest()
+
+
+
+def denest_key_fields(this_json, stream_name, path, dimensions_list):
+    """
+    Denest keys values list to dimension_list keys
+    """
+    new_json = this_json
+    idx = 0
+    for record in list(new_json[path]):
+        for key in list(record.keys()):
+            if isinstance(record[key], list) and key == "keys":
+                dim_num = 0
+                # Add dimensions_hash_key for performance_report_custom
+                if stream_name == "performance_report_custom":
+                    dims_md5 = str(hash_data(json.dumps(record[key], sort_keys=True)))
+                    new_json[path][idx]["dimensions_hash_key"] = dims_md5
+                for dimension in dimensions_list:
+                    new_json[path][idx][dimension] = record[key][dim_num]
+                    dim_num = dim_num + 1
+        idx = idx + 1
+    return new_json
+
+
+def add_site_url(this_json, path, site):
+    """
+    Adds site_url key, value to each object
+    """
+    new_json = this_json
```

Comment:
Are you intending to avoid changing the value of passed `this_json` parameter here ?

If so, then you may need to do something like :
```
from copy import deepcopy

new_json = deepcopy(this_json)

```
Otherwise it'll be a redundant statement and you could directly use `this_json` variable directly.

Reply:
Its redundant. so removed